File: Response_Handler/SupabaseHandler.py
from datetime import datetime
import os
from dotenv import load_dotenv
from supabase import Client, create_client

from Response_Handler.API_Models import Team

class SupabaseHandler:
    def __init__(self, supabase_url=None, supabase_key=None):
        if not supabase_url or not supabase_key:
            load_dotenv()
            supabase_url = os.getenv("SUPABASE_URL")
            supabase_key = os.getenv("SUPABASE_KEY")
            if not supabase_url or not supabase_key:
                raise RuntimeError("SUPABASE_URL and SUPABASE_KEY must be set in .env")
        
        self.supabase: Client = create_client(supabase_url, supabase_key)
        self.table = "season_2024"
        self.team_data = {}

    def get_response(self, team_number, table_name="null"):
        """
        Fetches data from the Supabase table based on the provided query.

        Args:
            table_name (str): The name of the table to query.
            query (dict): The query parameters.

        Returns:
            list: A list of records matching the query.
        """
        if table_name == "null":
            table_name = f"season{self.find_year()}"
        if table_name not in self.team_data:
            self.team_data[table_name] = {}

            raw_data = self.supabase.table(self.table).select("*").execute().data or []
            for row in raw_data:
                tnum = row.get("teamNumber")
                if tnum is not None:
                    self.team_data[table_name][tnum] = Team(
                        teamName=row["teamName"],
                        sponsors=row["sponsors"],
                        location=row["location"],
                        teamNumber=tnum,
                        autoOPR=row["autoOPR"],
                        teleOPR=row["teleOPR"],
                        endgameOPR=row["endgameOPR"],
                        overallOPR=row["overallOPR"],
                        penalties=row["penalties"],
                        autoRank=row.get("autoRank"),
                        teleRank=row.get("teleRank"),
                        endgameRank=row.get("endgameRank"),
                        overallRank=row.get("overallRank"),
                        penaltyRank=row.get("penaltyRank"),
                        profileUpdate=row.get("profileUpdate"),
                        eventDate=row.get("eventDate"),
                    )

        return self.team_data[table_name].get(team_number, Team())
    
    @staticmethod     
    def find_year():
        current_date = datetime.now()
        return current_date.year - 1 if current_date.month < 8 else current_date.year

# SupabaseHandler = SupabaseHandler()
# print(SupabaseHandler.get_response(14584))

File: src/supabase_handler.rs
use anyhow::{Context, Result};
use chrono::{Datelike, Utc};
use postgrest::Postgrest;
use serde::{Deserialize, Deserializer};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TeamData {
    #[serde(default, deserialize_with = "deserialize_i32_default")]
    pub team_number: i32,
    #[serde(default, deserialize_with = "deserialize_string_default")]
    pub team_name: String,
    pub sponsors: Option<String>,
    pub location: Option<String>,
    #[serde(default, deserialize_with = "deserialize_f64_default")]
    pub auto_opr: f64,
    #[serde(default, deserialize_with = "deserialize_f64_default")]
    pub tele_opr: f64,
    #[serde(default, deserialize_with = "deserialize_f64_default")]
    pub endgame_opr: f64,
    #[serde(default, deserialize_with = "deserialize_f64_default")]
    pub overall_opr: f64,
    #[serde(default, deserialize_with = "deserialize_f64_default")]
    pub penalties: f64,
    pub auto_rank: Option<i32>,
    pub tele_rank: Option<i32>,
    pub endgame_rank: Option<i32>,
    pub overall_rank: Option<i32>,
    pub penalty_rank: Option<i32>,
    pub profile_update: Option<String>,
    pub event_date: Option<String>,
}

fn deserialize_i32_default<'de, D>(deserializer: D) -> std::result::Result<i32, D::Error>
where
    D: Deserializer<'de>,
{
    let value: Option<serde_json::Value> = Option::deserialize(deserializer)?;
    let parsed = value
        .as_ref()
        .and_then(|v| v.as_i64().or_else(|| v.as_str().and_then(|s| s.parse::<i64>().ok())))
        .unwrap_or(0);

    Ok(parsed as i32)
}

fn deserialize_f64_default<'de, D>(deserializer: D) -> std::result::Result<f64, D::Error>
where
    D: Deserializer<'de>,
{
    let value: Option<serde_json::Value> = Option::deserialize(deserializer)?;
    Ok(value
        .as_ref()
        .and_then(|v| v.as_f64().or_else(|| v.as_str().and_then(|s| s.parse::<f64>().ok())))
        .unwrap_or(0.0))
}

fn deserialize_string_default<'de, D>(deserializer: D) -> std::result::Result<String, D::Error>
where
    D: Deserializer<'de>,
{
    let value: Option<serde_json::Value> = Option::deserialize(deserializer)?;
    Ok(value
        .as_ref()
        .and_then(|v| v.as_str().map(ToOwned::to_owned))
        .unwrap_or_else(|| "Unknown".to_owned()))
}

pub struct SupabaseHandler {
    supabase_url: String,
    supabase_key: String,
    cache: Arc<RwLock<HashMap<i32, TeamData>>>,
    table_name: String,
}

impl SupabaseHandler {
    pub fn new(supabase_url: String, supabase_key: String) -> Self {
        let year = Self::find_year();
        let table_name = format!("season_{}", year);

        Self {
            supabase_url,
            supabase_key,
            cache: Arc::new(RwLock::new(HashMap::new())),
            table_name,
        }
    }

    pub fn find_year() -> i32 {
        let now = Utc::now();
        let year = now.year();
        let month = now.month();

        if month < 8 { year - 1 } else { year }
    }

    pub async fn get_team(&self, team_number: i32) -> Result<TeamData> {
        {
            let cache = self.cache.read().unwrap();
            if let Some(team) = cache.get(&team_number) {
                return Ok(team.clone());
            }
        }

        // Create postgrest client with Supabase authentication
        let client = Postgrest::new(format!("{}/rest/v1", self.supabase_url))
            .insert_header("apikey", &self.supabase_key)
            .insert_header("Authorization", format!("Bearer {}", self.supabase_key));

        let response = client
            .from(&self.table_name)
            .select("*")
            .eq("teamNumber", &team_number.to_string())
            .limit(1)
            .execute()
            .await
            .context("Failed to query Supabase")?;

        let body = response
            .text()
            .await
            .context("Failed to read response body")?;
        let mut teams: Vec<TeamData> =
            serde_json::from_str(&body).with_context(|| {
                let snippet: String = body.chars().take(300).collect();
                format!("Failed to parse team data. Response starts with: {}", snippet)
            })?;

        let team = teams.pop().context("Team not found")?;

        {
            let mut cache = self.cache.write().unwrap();
            cache.insert(team_number, team.clone());
        }

        Ok(team)
    }
}
